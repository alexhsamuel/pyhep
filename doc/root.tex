\section{\module{hep.root} -- Root interface}

\declaremodule{extension}{hep.root}
\modulesynopsis{Root interface}

The \module{hep.root} module provides access from Python to features in
Root, a framework for high energy physics data management and analysis.
The applicable components of Root are included with \pyhep and need not
be provided separately.

\subsection{Root files}

Root includes an implementation of histograms and ntuples (which are
called ``trees'' in Root).  It implements a file format for storing
these in structured disk files.  \pyhep provides read and write access
to histograms and trees stored in Root files.

The \function{open} function access a Root file:

\begin{funcdesc}{open}{path\optional{, mode="r", purge_cycles=1}}
 Create or open a Root file at \var{path}.  

 \var{mode} is the access mode, analogous to Python's built-in
 \function{file} and \function{open} functions.  If \code{"r"}, opens an
 existing file for reading.  If \code{"r+"}, \code{"w"}, \code{"a"}, or
 \code{"a+"}, opens an existing file for reading and writing, or creates
 the file if it doesn't exist.  If \code{"w+"}, creates a file or
 overwrites an existing file.

 If \var{purge_cycles} is set to a true value and the file is open for
 writing, the entire file will be purged when it is closed.  
\end{funcdesc}

The return value from \function{open} is a \class{File} object.  The
file remains open until it is deleted.  

In \pyhep, all paths in Root files are specified relative to the root
directory of the Root file.  The root directory itself is represented by
an empty string, and the path separator is a forward slash.  Note that
there is no notion of working directory or relative paths.

Root files also support revision cycles; see the Root documentation for
details.  Unless \function{root.open} is invoked with
\code{purge_cycles=0} (or the file is opened read-only), \pyhep
``purges'' the entire file when the file is closed.  This removes cycles
except the most recent for each object in the file, whether or not the
object was written or revised in the current session.

\pyhep stores additional information along with tables and histograms in
Root files.  This information, called \emph{metadata}, consists of
useful information about histograms and tables that is not handled
directly by the Root file format.  \pyhep stores metadata in
subdirectories named ``_pyhep_metadata'' in the Root file.  These
subdirectories and their contents may be removed safely without injuring
the primary data in the histograms and tables they are associated with,
but doing so will delete configuration and annotations accessible from
\pyhep.

\subsubsection{\class{File} objects}

A \class{File} object has the following methods and attributes:

\begin{memberdesc}{writeable}
 True if the file is open for writing as well as reading.
\end{memberdesc}

\begin{methoddesc}{listdir}{\optional{path}}
 Return the contents of the directory specified by \var{path}.  The
 return value is a sequence of \code{(name, title, class_name)} tuples,
 where \var{name} is the object's name, \var{title} is the object's
 title, and \var{class_name} is the object's class name.
\end{methoddesc}

\begin{methoddesc}{mkdir}{path\optional{, title}}
 Create a new directory at \var{path}.  Optionally, a title may be
 specified in the \var{title} argument.
\end{methoddesc}

\begin{methoddesc}{rmdir}{path}
 Remove the directory at \var{path}.
\end{methoddesc}

\begin{methoddesc}{save}{path, object}
 Store a histogram \var{object} at \var{path}.  One- and two-dimensional
 histograms are supported.  
\end{methoddesc}

\begin{methoddesc}{load}{path}
 Load a histogram or connect to an ntuple at \var{path}.  If \var{path}
 is a one- or two-dimensional histogram, returns a histogram object with
 its contents.  If \var{path} is a tree, returns a table object (see
 \module{hep.table}) accessing it.
\end{methoddesc}

\begin{methoddesc}{createTable}{path, schema\optional{, title,
separate_branches=0, branch_name="default"}}
 Create a table as a tree at \var{path} in the Root file, using
 columns from \var{schema}, a \class{hep.table.Schema} instance.  

 By default, column values is stored in a leaves grouped in a single
 branch, named "default".  Another name may be specified for the branch
 with the \var{branch_name} argument.  Alternately, if
 \var{separate_branches} is true, each leaf is placed in a separate
 branch, whose name is the column name.

 These column types are supported: \constant{"int8"},
 \constant{"int16"}, \constant{"int32"}, \constant{"float32"}, and
 \constant{"float64"}.
\end{methoddesc}

\begin{methoddesc}{remove}{path}
 Remove the tree or histogram at \var{path}.  Use \method{rmdir} to
 remove a directory.
\end{methoddesc}

\begin{methoddesc}{normpath}{path}
 Return the canonical form of absolute path \var{path} in the Root
 file.
\end{methoddesc}

\begin{methoddesc}{join}{*components}
 Join together \var{components} to form an absolute path.
\end{methoddesc}

\begin{methoddesc}{split}{path}
 Return \code{(dir, name)}, where \var{dir} is the portion of \var{path}
 up to the last directory separator, and \var{name} is the remainder.
\end{methoddesc}

