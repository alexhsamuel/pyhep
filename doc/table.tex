\section{\module{hep.table} --- Tables}

\declaremodule{extension}{hep.table}
\modulesynopsis{Tables.}

A \pyhep \emph{table} is a flat database table with columns of fixed
type and storage size.  Tables are append-only: new rows may be added to
the end of the table, but rows may not be inserted elsewhere, nor can
rows be modified or deleted.  

Each table has a fixed \emph{schema}, which determines the data items
stored in the table and their types.  The schema includes an unordered
set of \emph{columns}, each identified by a name.  Each column has a
fixed data type.  A schema may also include \emph{expressions}, which
are derived quantities which are computed for each row from values of
columns and other expressions for that row.  

The table's data is an ordered sequence of \emph{rows}, each of which
contains exactly one value for each column in the table's schema.

Multiple table implementations are provided.  Implementations provide
different interfaces for creating and opening tables.  However, objects
representing tables and rows satisfy the same protocol across all
implementations.


\subsection{Schema objects}

An instance of \class{hep.table.Schema} represents a table schema.  

\begin{memberdesc}{columns}
 \readonly A sequence of columns in the schema.  Each element is a
 \class{Column} instance.  The order of the columns is unspecified.
\end{memberdesc}

\begin{memberdesc}{expressions}
 \readonly A sequence of expressions in the schema.  Each element is
 expression object (see \module{hep.expr}).  The order of the
 expressions is unspecified.
\end{memberdesc}

\begin{methoddesc}{addColumn}{name, type\optional{, **attributes}}
 Add a column to the sequence.  \var{name} is a character string, and
 \var{type} is a column type (see below).  If additional keyword
 arguments are provided, they are added to the column as attributes.
 New columns should not be added to a schema after the schema has been
 used to create a table.
\end{methoddesc}

\begin{methoddesc}{addExpression}{name, expression\optional{, **attributes}}
 Add an expression to the sequence.  \var{name} is a character string,
 and \var{expression} is an expression as a string or expression object.
 If additional keyword arguments are provided, they are added to the
 expression object as attributes.
\end{methoddesc}

An instance of \class{hep.table.Column} describes a column in a schema.
Do not instantiate this class directly; instead, use the
\method{addColumn} memthod of class \class{Schema}.

\begin{memberdesc}{name}
 \readonly The column's name.
\end{memberdesc}

\begin{memberdesc}{type}
 \readonly The column's data type.
\end{memberdesc}

The column's data type is specified by one of these names:
\begin{tableiii}{ccc}{code}{name}{C type}{Python Type}
  \lineiii{"int8"}{\code{signed char}}{\code{int}}
  \lineiii{"int16"}{\code{signed short}}{\code{int}}
  \lineiii{"int32"}{\code{signed int}}{\code{int}}
  \lineiii{"float32"}{\code{float}}{\code{float}}
  \lineiii{"float64"}{\code{double}}{\code{float}}
\end{tableiii}


An instance of \class{hep.table.Expression} describes a column in a
schema.  Do not instantiate this class directly; instead, use the
\method{addExpression} memthod of class \class{Schema}.

\begin{memberdesc}{name}
 \readonly The expressions's name.
\end{memberdesc}

\begin{memberdesc}{expression}
 \readonly An \class{Expression} instance; see \module{hep.expr}.
\end{memberdesc}

\begin{memberdesc}{formula}
 \readonly A string representation of the expression.
\end{memberdesc}



\subsection{Table objects}

A table object represents an in-memory table or a connection to an
externally-stored table.  In the former case, the table is erased when
the table object is deleted.  In the latter case, the connection is
closed but the table persists when the table object is deleted.

A table object satisfies the Python read-only sequence protocol.  The
sequence elements are the rows of the sequence.  Thus, the
\function{len} function returns the number of rows in the table, and
the \method{__getitem__} method extracts a row by its position in the
sequence. 

In addition, these methods are provided:

\begin{methoddesc}{append}{row\optional{, **kw_args}}
 (Read-write tables only.)  Appends a row to the end of the table.  The
 \var{row} must is a dictionary or other mapping object which maps
 column names to values.  Additional column values may be specified as
 keyword arguments.  If a column is specified both in \var{row} and in
 \var{**kw_args}, the value from latter is used.  A value must be
 specified for each column, and no other names may be specified.
 Returns the row index of the new row.
\end{methoddesc}

\begin{methoddesc}{__iter__}{}
 Returns an iterator over all rows in the table.
\end{methoddesc}

\begin{memberdesc}{rows}
 An interator over all rows in the table.
\end{memberdesc}

\begin{methoddesc}{select}{expr}
 Returns an iterator over rows in the table for which expression
 \var{expr} is true.  \var{expr} may be a string expression formula or
 an expression object.
\end{methoddesc}


\subsection{Row objects}

A row object may represent a row in a table, or a row that is compatible
with a table's schema (but not one of the rows in that table).

The row object satisfies the Python map protocol.  Each key is the name
of a column in the table's schema.  The corresponding map value is the
row's value for that column.  Row objects should be treated as
read-only; it is not possible to modify a table row by changing values
in a row object.

In addition, a row object provides these members and methods:

\begin{methoddesc}{asDict}{}
 Return a new Python dictionary of the column names and corresponding
 values in the row.
\end{methoddesc}

\begin{memberdesc}{index}
 \readonly The index of the row in the table.  The value of this
 attribute is undefined if the row was obtained from the table's
 \method{newRow} method and has not yet been passed to \method{append}.
\end{memberdesc}

\begin{memberdesc}{table}
 \readonly The table to which the row belongs.
\end{memberdesc}


\subsection{Standard implementation}

\pyhep's standard table implementation, in \module{hep.table}, uses a
simple, binary, on-disk representation of the table's data.  Table rows
are automatically loaded from disk as needed.

The \module{hep.table} module provides functions \function{create} and
\function{open} to create and open new tables, respectively.

\begin{funcdesc}{create}{filename, schema}
 Create a new table.  The table is stored in a file named by
 \var{filename}, which is created or overwritten.  The \var{schema}
 argument is a map from column names to \class{Column} instances,
 specifying the columns in the table.

 The return value is a table object, which is opened to the new, empty
 in write mode.
\end{funcdesc}

\begin{funcdesc}{open}{filename\optional{, mode="r"}}
 Open an existing table stored in the file named by \var{filename}.  The
 \var{mode} argument specifies the mode in which to open the table:
 \constant{"r"} to open the table in read-only mode, or \constant{"w"}
 to open the table in read-write mode.

 The return value is a table object.
\end{funcdesc}

