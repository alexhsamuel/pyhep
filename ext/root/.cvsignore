root.py
root_wrap.cc
root_wrap.doc
*Dict.cc
*Dict.h
*.pyc
depend.mk
ROOTSYS
