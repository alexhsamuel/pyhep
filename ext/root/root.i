//------------------------------------------------------------*- c++ -*-
//
// root.i
//
// Copyright 2003 by Alex Samuel.  All rights reserved.
//
//----------------------------------------------------------------------

// SWIG interface file for Root libraries.

// -*- c++ -*-

%module root
%{
#include "TROOT.h"
#include "TKey.h"
#include "TDirectory.h"
#include "TFile.h"
#include "TLeaf.h"
#include "TLeafD.h"
#include "TLeafF.h"
#include "TLeafI.h"
#include "TList.h"
#include "TObject.h"
#include "TTree.h"
#include "TH1.h"
#include "TH2.h"
#include "TH3.h"
#include "TAxis.h"

class IndexError 
{
public:
  IndexError(int index) : index_(index) {};
  const int index_;
};

class StopIteration {};

%}

%inline %{
long 
getPointerValue(void* ptr) 
{
   return (long) ptr;
}
%}

%except(python) {
  try {
    $function
  }
  catch (IndexError e) {
    PyErr_Format(PyExc_IndexError, "%d", e.index_);
    return NULL;
  }
  catch (StopIteration) {
    PyErr_SetString(PyExc_StopIteration, "end of iterator");
    return NULL;
  }
}


class TObject
{
public:

  bool IsZombie() const;
  virtual const char* ClassName() const;
  virtual const char* GetName() const;
  virtual const char* GetTitle() const;
  virtual bool InheritsFrom(const char *classname) const;
  virtual void Print(const char* option) const;
  virtual int Write(const char* name=0, int option=0, int bufsize=0);

};


class TNamed
  : public TObject
{
public:

  virtual void SetName(const char* name);
  virtual void SetTitle(const char* title);

};


%pragma no_default

class TIterator
  : public TObject
{
public:

  virtual TObject* Next();

};

%addmethods TIterator {
  TIterator* __iter__() 
  {
    return self;
  };


  TObject* next()
  {
    TObject* object = self->Next();
    if (object == NULL)
      throw StopIteration();
    else
      return object;
  };
};



class TClass
  : public TObject
{
public:

  bool InheritsFrom(const char* cl) const;

};


class TKey
  : public TNamed
{
public:

  const char* GetClassName() const;
  short GetCycle() const;
  virtual TClass* IsA() const;

};


class TDirectory
  : public TNamed
{
public:

  virtual bool cd();
  virtual void Delete(const char* namecycle);
  virtual TObject* Get(const char* namecycle);
  TFile* GetFile() const;
  TKey* GetKey(const char* name, short cycle=9999);
  TList* GetList() const;
  TList* GetListOfKeys() const;
  TObject* GetMother() const;
  const char* GetPath() const;
  virtual TDirectory* mkdir(const char* name, const char* title);
  void Purge(short nKeep);
  
};


class TFile
  : public TDirectory
{
public:

  TFile(const char* fname, 
	char* option="", 
	const char* ftitle="", 
	int compress=1);

  virtual bool cd(const char *path = 0);
  virtual void Close(char* option="");
  virtual int Write(const char* name=0, int opt=0, int bufsiz=0);

};


class TList
  : public TObject
{
public:

  virtual int GetSize() const;
  virtual TObject* At(int idx) const;

};


%addmethods TList {
  int __len__() 
  {
    return self->GetSize();
  };


  TObject* __getitem__(int idx) 
  {
    if (idx < 0 || idx >= self->GetSize()) 
      throw IndexError(idx);
    else
      return self->At(idx);
  };
};



class TTree
  : public TObject
{
public:

  virtual int GetEntry(int entry=0, int getall=0);
  virtual double GetEntries() const;
  virtual TIterator* GetIteratorOnAllLeaves();
  virtual void Print();

};


class TLeaf
  : public TObject
{
public:
  
  virtual void SetAddress(void* add=0);

};

%addmethods TLeaf {
  void SetAddressInt(int add=0) 
  {
    self->SetAddress((void*) add);
  };
};


class TLeafD
  : public TLeaf
{
};


class TLeafF
  : public TLeaf
{
};


class TLeafI
  : public TLeaf
{
};


class TH1
  : public TNamed
{
public:

  virtual TAxis* GetXaxis();
  virtual TAxis* GetYaxis();
  virtual TAxis* GetZaxis();
  virtual double GetBinContent(int bin);
  virtual double GetBinError(int bin);
  virtual double GetEntries();
  virtual void SetBinContent(int bin, double content);
  virtual void SetBinError(int bin, double error);
  virtual void SetEntries(double entries);

};


class TH1F 
  : public TH1
{
public:

  TH1F(const char* name, const char* title, int nbinsx, 
       double xlow, double xup);

};


class TH1D
  : public TH1
{
public:

  TH1D(const char* name, const char* title, int nbinsx, 
       double xlow, double xup);

};


class TH1C
  : public TH1
{
public:

  TH1C(const char* name, const char* title, int nbinsx, 
       double xlow, double xup);

};


class TH1S
  : public TH1
{
public:

  TH1S(const char* name, const char* title, int nbinsx, 
       double xlow, double xup);

};


class TH2
  : public TH1
{
public:

  virtual double GetBinContent(int binx, int biny);
  virtual double GetBinError(int binx, int biny);
  virtual void SetBinContent(int binx, int biny, double content);
  virtual void SetBinError(int binx, int biny, double error);

};


class TH2C
  : public TH2
{
public:

  TH2C(const char* name, const char* title, int nbinsx, double xlow,
       double xup, int nbinsy, double ylow, double yup);

};


class TH2D
  : public TH2
{
public:

  TH2D(const char* name, const char* title, int nbinsx, double xlow,
       double xup, int nbinsy, double ylow, double yup);

};


class TH2F
  : public TH2
{
public:

  TH2F(const char* name, const char* title, int nbinsx, double xlow,
       double xup, int nbinsy, double ylow, double yup);

};


class TH2S
  : public TH2
{
public:

  TH2S(const char* name, const char* title, int nbinsx, double xlow,
       double xup, int nbinsy, double ylow, double yup);

};


class TH3
  : public TH1
{
};


class TAxis
  : public TObject
{
public:

  int GetNbins();
  double GetXmin();
  double GetXmax();

};
