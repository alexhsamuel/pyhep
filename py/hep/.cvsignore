*.pyc
*.pyo
config.py
*-install
pdt.pickle
