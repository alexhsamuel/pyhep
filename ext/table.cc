//----------------------------------------------------------------------
//
// table.cc
//
// Copyright 2003 by Alex Samuel.  All rights reserved.
//
//----------------------------------------------------------------------

//----------------------------------------------------------------------
// includes
//----------------------------------------------------------------------

#include <cassert>
#include <cerrno>
#include <complex>
#include <fcntl.h>
#include <libgen.h>
#include <stdint.h>
#include <sys/stat.h>
#include <sys/types.h>

#include "table.hh"
#include "value.hh"

namespace table {

namespace {

//----------------------------------------------------------------------
// constants
//----------------------------------------------------------------------

const size_t
type_sizes[] = { 
  1,    // TYPE_BOOL
  1,    // TYPE_INT_8
  2,    // TYPE_INT_16
  4,    // TYPE_INT_32
  4,    // TYPE_FLOAT_32
  8,    // TYPE_FLOAT_64
  8,    // TYPE_COMPLEX_64
  16,   // TYPE_COMPLEX_128
 };


const char* const
type_names[] = {
  "bool",       // TYPE_BOOL
  "int8",       // TYPE_INT_8
  "int16",      // TYPE_INT_16
  "int32",      // TYPE_INT_32
  "float32",    // TYPE_FLOAT_32
  "float64",    // TYPE_FLOAT_64
  "complex64",  // TYPE_COMPLEX_64
  "complex128", // TYPE_COMPLEX_128
};


const int
file_format_magic_number = 0x11a66826;

const int
file_format_version_number = 6;


//----------------------------------------------------------------------
// private types
//----------------------------------------------------------------------

// We have stdint.h to provide int32_t etc. but nothing for similar
// floating-point types.  Define the ones we need here.

// FIXME: These need to be configured specifically for each
// architecture. 

typedef float
float32_t;

typedef double
float64_t;

typedef std::complex<float>
complex64_t;

typedef std::complex<double>
complex128_t;

//----------------------------------------------------------------------

struct FileFormatHeader
{
  /* A magic number, identifying this file as a table file.  */
  int magic_number_;

  /* The file format version number.  */
  int version_number_;

  /* The number of rows in the table.  */
  int num_rows_;

  /* The offset to the data for the first row.  */
  off64_t first_row_offset_;

  /* Ignored, but retained for file format compatibility.  */
  off64_t metadata_offset_;

  /* Ignored, but retained for file format compatibility.  */
  off64_t metadata_len_;
};


//----------------------------------------------------------------------

class Buffer 
{
public:

  Buffer(size_t size) : pointer_(new char[size]) {}
  ~Buffer() { delete[] pointer_; }

  char* const pointer_;

};


//----------------------------------------------------------------------
// helper functions
//----------------------------------------------------------------------

inline size_t
xwrite(int fd,
       const void* buffer,
       size_t count)
  throw (FileError)
{
  size_t result = ::write(fd, buffer, count);
  if (result != count)
    throw FileError(strerror(errno));
  return count;
}


inline void
xread(int fd, 
      void* buffer,
      size_t count)
  throw (FileError)
{
  size_t result = ::read(fd, buffer, count);
  if (result != count)
    throw FileError(strerror(errno));
}


inline void
xseek(int fd,
      off64_t offset)
  throw (FileError)
{
  off64_t result = ::lseek64(fd, offset, SEEK_SET);
  if (result != offset)
    throw FileError(strerror(errno));
}


inline std::string
getMetadataPath(const std::string& path)
{
  return path + ".metadata";
}


size_t
writeSchema(int fd, 
	    const Schema* schema)
{
  assert(fd >= 0);
  assert(schema != NULL);
  size_t written = 0;

  size_t size = schema->getSize();
  written += xwrite(fd, &size, sizeof(size_t));

  int num_columns = schema->getNumColumns();
  written += xwrite(fd, &num_columns, sizeof(int));

  for (int c = 0; c < num_columns; ++c) {
    const Column& column = schema->getColumn(c);

    ColumnType type = column.getType();
    written += xwrite(fd, &type, sizeof(ColumnType));

    std::string name = column.getName();
    int name_length = name.length();
    const char* name_string = name.data();
    written += xwrite(fd, &name_length, sizeof(int));
    written += xwrite(fd, name_string, name_length);

    size_t offset = schema->getColumnOffset(c);
    written += xwrite(fd, &offset, sizeof(offset));
  }

  return written;
}


Schema*
readSchema(int fd)
{
  assert(fd >= 0);
  
  size_t size;
  xread(fd, &size, sizeof(size_t));

  CustomSchema* schema = new CustomSchema(size);

  int num_columns;
  xread(fd, &num_columns, sizeof(int));

  for (int c = 0; c < num_columns; ++c) {
    ColumnType type;
    xread(fd, &type, sizeof(ColumnType));

    int name_length;
    xread(fd, &name_length, sizeof(int));
    char name[name_length + 1];
    xread(fd, name, name_length);
    name[name_length] = '\0';

    size_t offset;
    xread(fd, &offset, sizeof(offset));

    schema->addColumn(Column(name, type), offset);
  }

  return schema;
}


}  // anonymous namespace


//----------------------------------------------------------------------
// class Row
//----------------------------------------------------------------------

Row::Row(const Schema* schema)
  : schema_(schema)
{
  assert(schema_ != 0);
  data_ = new char[schema_->getSize()];
}


Row::~Row()
{
  delete [] data_;
}


#define ROW_GET(TYPE, OFFSET) \
  (*((TYPE*) (data_ + (OFFSET))))

Value
Row::getValue(const int column_index)
  const
  throw (WrongColumnType)
{
  ColumnType type = schema_->getColumn(column_index).getType();
  size_t offset = schema_->getColumnOffset(column_index);
  switch (type) {
  case TYPE_NONE:
    abort();

  case TYPE_BOOL:
    return Value::make((bool) ROW_GET(int8_t, offset));
		       
  case TYPE_INT_8:
    return Value::make((long) ROW_GET(int8_t, offset));

  case TYPE_INT_16:
    return Value::make((long) ROW_GET(int16_t, offset));
    
  case TYPE_INT_32:
    return Value::make((long) ROW_GET(int32_t, offset));
    
  case TYPE_FLOAT_32:
    return Value::make((double) ROW_GET(float32_t, offset));

  case TYPE_FLOAT_64:
    return Value::make((double) ROW_GET(float64_t, offset));

  case TYPE_COMPLEX_64: 
    return Value::make((std::complex<double>) ROW_GET(complex64_t, offset));

  case TYPE_COMPLEX_128: 
    return Value::make((std::complex<double>) ROW_GET(complex128_t, offset));

  default:
    abort();
  }
}


#define ROW_SET(TYPE, OFFSET, VALUE) \
  *((TYPE*) (data_ + (OFFSET))) = ((TYPE) (VALUE))

void
Row::setValue(const int column_index,
	      const Value& value)
{
  assert(column_index >= 0 && column_index < schema_->getNumColumns());

  ColumnType type = schema_->getColumn(column_index).getType();
  size_t offset = schema_->getColumnOffset(column_index);
  switch (type) {
  case TYPE_NONE:
    abort();

  case TYPE_BOOL:
    ROW_SET(int8_t, offset, value.as_bool());
    break;
		       
  case TYPE_INT_8:
    ROW_SET(int8_t, offset, value.as_long());
    break;

  case TYPE_INT_16:
    ROW_SET(int16_t, offset, value.as_long());
    break;
    
  case TYPE_INT_32:
    ROW_SET(int32_t, offset, value.as_long());
    break;
    
  case TYPE_FLOAT_32:
    ROW_SET(float32_t, offset, value.as_double());
    break;

  case TYPE_FLOAT_64:
    ROW_SET(float64_t, offset, value.as_double());
    break;

  case TYPE_COMPLEX_64:
    ROW_SET(complex64_t, offset, value.as_complex());
    break;

  case TYPE_COMPLEX_128:
    ROW_SET(complex128_t, offset, value.as_complex());
    break;

  default:
    abort();
  }
}  


//----------------------------------------------------------------------
// class Schema
//----------------------------------------------------------------------

Schema::~Schema()
{
}


int
Schema::whichColumn(const std::string& name)
  const
  throw (NoColumn)
{
  int num_columns = columns_.size();
  for (int i = 0; i < num_columns; ++i) 
    if (columns_[i].column_.getName() == name)
      return i;

  throw NoColumn(name);
}


//----------------------------------------------------------------------
// class AutoSchema
//----------------------------------------------------------------------

AutoSchema::AutoSchema()
{
}


AutoSchema::~AutoSchema()
{
}


void
AutoSchema::addColumn(const Column& column)
{
  ColumnRecord record = { column, size_ };
  columns_.push_back(record);
  size_ += getTypeSize(column.getType());
}


//----------------------------------------------------------------------
// class Table
//----------------------------------------------------------------------

Table::~Table()
{
}


//----------------------------------------------------------------------
// class FileTable
//----------------------------------------------------------------------

FileTable*
FileTable::create(const Schema* schema,
		  const std::string& path,
		  mode_t mode)
{
  // Create the file.
  int fd = ::open64(path.c_str(), 
		    O_RDWR | O_CREAT | O_TRUNC | O_LARGEFILE, mode);
  if (fd < 0)
    throw FileError(strerror(errno));

  // Skip forward to leave enough room for the header, then write the
  // schema.
  xseek(fd, sizeof(FileFormatHeader));
  off64_t schema_size = writeSchema(fd, schema);

  // Construct the header info, now that we know how large the schema
  // is.
  FileFormatHeader header;
  header.magic_number_ = file_format_magic_number;
  header.version_number_ = file_format_version_number;
  header.num_rows_ = 0;
  header.first_row_offset_ = sizeof(FileFormatHeader) + schema_size;

  // Write the header at the beginning of the file.
  xseek(fd, 0);
  xwrite(fd, &header, sizeof(header));

  // Close the file.
  close(fd);

  // If there is a metadata file for this path, remove it.
  std::string metadata_path = getMetadataPath(path);
  if (::access(metadata_path.c_str(), X_OK))
    ::unlink(metadata_path.c_str());

  // Now open the newly-created table in the usual way.
  return new FileTable(path, O_RDWR);
}


FileTable*
FileTable::open(const std::string& path,
		const std::string& mode)
{
  int flags;
  if (mode == "r") 
    flags = O_RDONLY;
  else if (mode == "w") 
    flags = O_RDWR;
  else
    throw FileError("unknown mode");

  return new FileTable(path, flags);
}


FileTable::~FileTable()
{
  if (isWritable()) {
    // Build the header.
    FileFormatHeader header;
    header.magic_number_ = file_format_magic_number;
    header.version_number_ = file_format_version_number;
    header.num_rows_ = num_rows_;
    header.first_row_offset_ = first_row_offset_;

    // Write the header.
    xseek(fd_, 0);
    xwrite(fd_, &header, sizeof(header));

    int result = fsync(fd_);
    if (result != 0)
      throw FileError(strerror(errno));
  }

  int result;
  result = close(fd_);
  assert(result == 0);
}


void
FileTable::read(int64_t row_number, 
		Row* row)
{
  assert(row->getSchema() == this->getSchema());
  assert(row_number >= 0 && row_number < num_rows_);

  off64_t offset = first_row_offset_ + row_number * row_size_;
  xseek(fd_, offset);
  xread(fd_, row->getBuffer(), row_size_);
}


int
FileTable::append(const Row* row)
{
  assert(row->getSchema() == this->getSchema());

  if (! isWritable())
    throw NotWritable();

  int64_t row_number = num_rows_;
  off64_t offset = first_row_offset_ + row_number * row_size_;
  xseek(fd_, offset);
  xwrite(fd_, row->getBuffer(), row_size_);

  ++num_rows_;
  return row_number;
}


std::string
FileTable::getMetadata()
  const
{
  std::string path = getMetadataPath(path_);

  // Can we access the metadata file?
  if (::access(path.c_str(), R_OK) != 0)
    // Can't access the metadata file.  Either it doesn't exist or it's
    // inaccessible.  In either case, just return empty metadata.
    return "";

  // Get file information.
  struct stat file_info;
  if (::stat(path.c_str(), &file_info) != 0)
    return "";
  if (file_info.st_size == 0)
    // No metadata.
    return "";

  // Open the metadata file.
  int fd = ::open(path.c_str(), O_RDONLY);
  if (fd == -1) 
    return "";

  // Construct a buffer to hold the metadata.
  std::auto_ptr<Buffer> buffer(new Buffer(file_info.st_size));

  // Read the data.
  xread(fd, buffer->pointer_, file_info.st_size);
  // Clean up.
  ::close(fd);

  return std::string(buffer->pointer_, file_info.st_size);
}


void
FileTable::setMetadata(const std::string& data)
{
  std::string path = getMetadataPath(path_);

  if (data == "") {
    // We don't want a metadata file.  Remove one, if it exists.
    if (::access(path.c_str(), X_OK))
      ::unlink(path.c_str());
    // Don't write.
    return;
  }

  // Create the metadata file.
  int fd = ::open(path.c_str(), O_WRONLY | O_CREAT | O_TRUNC, 0666);
  if (fd == -1) 
    // Could not open the metadata file; fail silently.
    return;

  // Write the metadata.
  xwrite(fd, data.c_str(), data.length());
  // Clean up.
  close(fd);
}


FileTable::FileTable(const std::string& path,
		     int flags)
  : Table(NULL, flags == O_RDWR),
    path_(path)
{
  if (flags != O_RDWR && flags != O_RDONLY)
    // Invalid flags.
    assert(false);

  // Open the file.
  fd_ = ::open64(path.c_str(), flags | O_LARGEFILE);
  if (fd_ < 0) {
    throw FileError(strerror(errno));
  }

  // Read the header.
  FileFormatHeader header;
  xread(fd_, &header, sizeof(header));

  // Check the magic numbers.
  if (header.magic_number_ != file_format_magic_number)
    throw FileError("wrong file format");
  if (header.version_number_ != file_format_version_number)
    throw FileError("wrong file format version");

  // Extract other info from the header.
  num_rows_ = header.num_rows_;
  assert(num_rows_ >= 0);
  first_row_offset_ = header.first_row_offset_;

  // Read the schema.
  schema_ = readSchema(fd_);
  assert(schema_ != NULL);
  row_size_ = schema_->getSize();
}


//----------------------------------------------------------------------
// function definitions
//----------------------------------------------------------------------

size_t
getTypeSize(ColumnType type)
{
  return type_sizes[(int) type];
}


const char*
getTypeName(ColumnType type)
{
  int i = (int) type;
  if (i >= 0 && i < (int) TYPE_LAST)
    return type_names[i];
  else
    return NULL;
}


ColumnType
getTypeByName(const char* name)
{
  // Try to match the name.
  for (int i = 0; i < (int) TYPE_LAST; ++i)
    if (strcmp(name, type_names[i]) == 0)
      return (ColumnType) i;
  // No match.
  return TYPE_NONE;
}


//----------------------------------------------------------------------

}  // namespace Tables
