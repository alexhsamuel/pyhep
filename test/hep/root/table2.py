#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

import hep.root
import hep.table
from   hep.test import compare

#-----------------------------------------------------------------------
# test
#-----------------------------------------------------------------------

schema = hep.table.Schema()

root_file = hep.root.create("table2.root")
bar = root_file.mkdir("foo/bar")
table = hep.root.createTable("table", bar, schema, title="A nice test.")
for i in xrange(0, 10):
    table.append()
del root_file, bar, table

# Make sure the table is in the right directory.
table = hep.root.open("table2.root")["foo/bar/table"]
# Make sure an empty schema is handled correctly.
compare(len(table.schema.columns), 0)
compare(len(table), 10)
# Make sure the name and title are restored correctly.
compare(table.name, "table")
compare(table.title, "A nice test.")
