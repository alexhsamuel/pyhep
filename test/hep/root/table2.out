Traceback (most recent call last):
  File "table2.py", line 16, in ?
    bar = root_file.mkdir("foo/bar")
  File "/home/samuel/dev/pyhep/py/hep/fs.py", line 355, in mkdir
    subdir, key = self._splitdir(path, **kw_args)
  File "/home/samuel/dev/pyhep/py/hep/fs.py", line 477, in _splitdir
    subdir = subdir.setdefault(dir_key, {}, **kw_args)
  File "/home/samuel/dev/pyhep/py/hep/fs.py", line 379, in setdefault
    subdir.__set(key, default, **kw_args)
  File "/home/samuel/dev/pyhep/py/hep/fs.py", line 441, in __set
    self._set(key, value, **kw_args)
  File "/home/samuel/dev/pyhep/py/hep/root/__init__.py", line 226, in _set
    tkey = self.__getTKey(key)
  File "/home/samuel/dev/pyhep/py/hep/root/__init__.py", line 262, in __getTKey
    return _cast(key)
  File "/home/samuel/dev/pyhep/py/hep/root/__init__.py", line 600, in _cast
    parts = object.this.split("_")
AttributeError: 'PySwigObject' object has no attribute 'split'
